R/calcHDDCDD.R

```diff
@@ -1,101 +1,714 @@
-#' @importFrom magclass setNames setYears as.magpie getNames lowpass getYears mbind
-#' @importFrom dplyr %>%
-#' @importFrom quitte as.quitte
-#' @importFrom utils tail
-#' @export
-
-calcHDDCDD <- function(tlimit = 18) {
-
-  stopifnot(as.character(tlimit) %in% as.character(17:25))
-
-  data <- readSource("HDDCDD", subtype = tlimit)
-
-  # yearsScenario = all years for which history is either NA or 0
-  index <- apply(data[, , paste0("history.HDD.history.", tlimit)], c(2, 3),
-                 function(x) {
-                   return(all(is.na(x) | x == 0))
-                 })
-
-  yearsScenario <- c(gsub("y", "", attributes(index)$dimnames$period[index]) %>% as.integer(), 2100)
-  yearsHistory <- setdiff(getYears(data, as.integer = TRUE), yearsScenario)
-
-  convergenceYear <- 2030
-  convergenceYears <- c(max(yearsHistory), yearsScenario[yearsScenario <= convergenceYear])
-  afterConvergenceYears <- yearsScenario[!yearsScenario %in% convergenceYears]
-  lambda <- c(rep(1, length(yearsHistory)),
-              tail(seq(1, 0, length.out = length(convergenceYears)), -1),
-              rep(0, length(afterConvergenceYears)))
-  names(lambda) <- c(yearsHistory, yearsScenario)
-  lambda <- as.magpie(lambda)
-
-  rcps     <- grep("history", getNames(data, TRUE)$rcp,      value = TRUE, invert = TRUE)
-  ssps     <- grep("history", getNames(data, TRUE)$scenario, value = TRUE, invert = TRUE)
-  variable <- grep("history", getNames(data, TRUE)$variable, value = TRUE, invert = TRUE)
-
-  data[is.na(data)] <- 0
-  data <- mbind(data, (1 / 3) * (setYears(data[, 2099, ], 2100) +
-                                 setYears(data[, 2098, ], 2100) +
-                                 setYears(data[, 2097, ], 2100))) # give a point for 2100
-
-  # distribute the history between all scenarios
-  data <- do.call("mbind", lapply(rcps, function(rcp) {
-    do.call("mbind", lapply(ssps, function(ssp) {
-      tmp <- data[, , rcp][, , ssp]
-      tmp[, yearsHistory, ] <- setNames(data[, yearsHistory, "history"][, , "history"],
-                                        getNames(tmp[, yearsHistory, ]))
+#' @author Robin Krekeler, Hagen Tockhorn
+#' @examples
+#'
+#' \dontrun{
+#' calcHDDCDD()
+#' }
+#'
+#' @importFrom madrat getConfig
+#' @importFrom raster cellStats
+#' @importFrom ncdf4 nc_open
+#' @importFrom tidyr %>%
+#' @importFrom dplyr mutate
+#' @importFrom rlang .data
+#' @importFrom pracma integral2
+#' @importFrom stringr str_sub
+#' @importFrom terra app nlyr tapp subset rast classify time
+
+
+calcHDDCDD <- function(mappingFile, bait=FALSE, multiscen = FALSE) {
+
+
+  # FUNCTIONS-------------------------------------------------------------------
+
+  # fill dates for unnamed data
+  #   ---> was supposed to go to convertISIMIP, doesn't work with country check
+  fillDates <- function(r, filename, bait=FALSE) {
+    if (grepl(".nc|.nc4", filename)) {
+      filename <- gsub(".nc|.nc4", "", filename)
+    }
+
+    if (bait) {
```

Comment:
General consideration: If you have if-clauses based on arguments set by the user, you have to expect unintended input. What if `bait` is not a boolean (e.g. `NULL`). In this case, You would get an error here that likely cannot be understood without looking into he code. My suggestion below won't fail and only return `TRUE`, is `bait` is `TRUE`. Any other input will lead to the else-clause being active. This is more robust but maybe you want the code to fail with wrong input. Just that you are aware.
```suggestion
    if (isTRUE(bait)) {
```